// @hint/hint-axe/dist/src/meta/aria.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: utils_types_1.Category.accessibility,
        description: i18n_import_1.getMessage('aria_description', 'en'),
        name: i18n_import_1.getMessage('aria_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('aria_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('aria_name', language);
    },
    id: 'axe/aria',
    schema: [
        {
            additionalProperties: false,
            properties: {
                'aria-allowed-attr': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-allowed-role': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-command-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-dialog-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-hidden-body': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-input-field-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-meter-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-progressbar-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-required-attr': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-required-children': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-required-parent': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-roledescription': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-roles': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-text': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-toggle-field-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-tooltip-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-treeitem-name': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-valid-attr': { enum: ['off', 'warning', 'error'], type: 'string' },
                'aria-valid-attr-value': { enum: ['off', 'warning', 'error'], type: 'string' },
                'empty-table-header': { enum: ['off', 'warning', 'error'], type: 'string' },
                'presentation-role-conflict': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        },
        {
            items: {
                enum: ['aria-allowed-attr', 'aria-allowed-role', 'aria-command-name', 'aria-dialog-name', 'aria-hidden-body', 'aria-input-field-name', 'aria-meter-name', 'aria-progressbar-name', 'aria-required-attr', 'aria-required-children', 'aria-required-parent', 'aria-roledescription', 'aria-roles', 'aria-text', 'aria-toggle-field-name', 'aria-tooltip-name', 'aria-treeitem-name', 'aria-valid-attr', 'aria-valid-attr-value', 'empty-table-header', 'presentation-role-conflict'],
                type: 'string'
            },
            typeof: 'array',
            uniqueItems: true
        }
    ],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
