// @hint/hint-axe/dist/src/meta/color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: utils_types_1.Category.accessibility,
        description: i18n_import_1.getMessage('color_description', 'en'),
        name: i18n_import_1.getMessage('color_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('color_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('color_name', language);
    },
    id: 'axe/color',
    schema: [
        {
            additionalProperties: false,
            properties: {
                'color-contrast': { enum: ['off', 'warning', 'error'], type: 'string' },
                'link-in-text-block': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        },
        {
            items: {
                enum: ['color-contrast', 'link-in-text-block'],
                type: 'string'
            },
            typeof: 'array',
            uniqueItems: true
        }
    ],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
