// @hint/hint-axe/dist/src/meta/sensory-and-visual-cues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: utils_types_1.Category.accessibility,
        description: i18n_import_1.getMessage('sensoryAndVisualCues_description', 'en'),
        name: i18n_import_1.getMessage('sensoryAndVisualCues_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('sensoryAndVisualCues_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('sensoryAndVisualCues_name', language);
    },
    id: 'axe/sensory-and-visual-cues',
    schema: [
        {
            additionalProperties: false,
            properties: {
                'meta-viewport': { enum: ['off', 'warning', 'error'], type: 'string' },
                'meta-viewport-large': { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        },
        {
            items: {
                enum: ['meta-viewport', 'meta-viewport-large'],
                type: 'string'
            },
            typeof: 'array',
            uniqueItems: true
        }
    ],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
