// @hint/hint-axe/dist/src/meta/structure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const i18n_import_1 = require("../i18n.import");
const meta = {
    docs: {
        category: utils_types_1.Category.accessibility,
        description: i18n_import_1.getMessage('structure_description', 'en'),
        name: i18n_import_1.getMessage('structure_name', 'en')
    },
    getDescription(language) {
        return i18n_import_1.getMessage('structure_description', language);
    },
    getName(language) {
        return i18n_import_1.getMessage('structure_name', language);
    },
    id: 'axe/structure',
    schema: [
        {
            additionalProperties: false,
            properties: {
                'avoid-inline-spacing': { enum: ['off', 'warning', 'error'], type: 'string' },
                'css-orientation-lock': { enum: ['off', 'warning', 'error'], type: 'string' },
                'definition-list': { enum: ['off', 'warning', 'error'], type: 'string' },
                dlitem: { enum: ['off', 'warning', 'error'], type: 'string' },
                'frame-tested': { enum: ['off', 'warning', 'error'], type: 'string' },
                'hidden-content': { enum: ['off', 'warning', 'error'], type: 'string' },
                list: { enum: ['off', 'warning', 'error'], type: 'string' },
                listitem: { enum: ['off', 'warning', 'error'], type: 'string' }
            }
        },
        {
            items: {
                enum: ['avoid-inline-spacing', 'css-orientation-lock', 'definition-list', 'dlitem', 'frame-tested', 'hidden-content', 'list', 'listitem'],
                type: 'string'
            },
            typeof: 'array',
            uniqueItems: true
        }
    ],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
